kmk/extensions/layers.py

```diff
@@ -0,0 +1,212 @@
+'''One layer isn't enough. Adds keys to get to more of them'''
+from micropython import const
+
+from kmk.extensions import Extension
+from kmk.key_validators import layer_key_validator
+from kmk.keys import make_argumented_key
+from kmk.kmktime import accurate_ticks, accurate_ticks_diff
+
+
+class LayerType:
+    '''Defines layer type values for readability'''
+
+    MO = const(0)
+    DF = const(1)
+    LM = const(2)
+    LT = const(3)
+    TG = const(4)
+    TT = const(5)
+
+
+class Layers(Extension):
+    '''Gives access to the keys used to enable the layer system'''
+
+    def __init__(self):
+        # Layers
+        make_argumented_key(
+            validator=layer_key_validator,
+            names=('MO',),
+            on_press=self._mo_pressed,
+            on_release=self._mo_released,
+        )
+        make_argumented_key(
+            validator=layer_key_validator, names=('DF',), on_press=self._df_pressed
+        )
+        make_argumented_key(
+            validator=layer_key_validator,
+            names=('LM',),
+            on_press=self._lm_pressed,
+            on_release=self._lm_released,
+        )
+        make_argumented_key(
+            validator=layer_key_validator,
+            names=('LT',),
+            on_press=self._lt_pressed,
+            on_release=self._lt_released,
+        )
+        make_argumented_key(
+            validator=layer_key_validator, names=('TG',), on_press=self._tg_pressed
+        )
+        make_argumented_key(
+            validator=layer_key_validator, names=('TO',), on_press=self._to_pressed
+        )
+        make_argumented_key(
+            validator=layer_key_validator,
+            names=('TT',),
+            on_press=self._tt_pressed,
+            on_release=self._tt_released,
+        )
+
+    start_time = {
+        LayerType.LT: None,
+        LayerType.TG: None,
+        LayerType.TT: None,
+        LayerType.LM: None,
+    }
+
+    def on_runtime_enable(self, keyboard):
+        return
+
+    def on_runtime_disable(self, keyboard):
+        return
+
+    def during_bootup(self, keyboard):
+        return
+
+    def before_matrix_scan(self, keyboard):
+        return
+
+    def after_matrix_scan(self, keyboard, matrix_update):
+        return
+
+    def before_hid_send(self, keyboard):
+        return
+
+    def after_hid_send(self, keyboard):
+        return
+
+    def on_powersave_enable(self, keyboard):
+        return
+
+    def on_powersave_disable(self, keyboard):
+        return
+
+    @staticmethod
+    def _df_pressed(key, state, *args, **kwargs):
+        '''
+        Switches the default layer
+        '''
+        state.active_layers[-1] = key.meta.layer
+        return state
+
+    @staticmethod
+    def _mo_pressed(key, state, *args, **kwargs):
+        '''
+        Momentarily activates layer, switches off when you let go
+        '''
+        state.active_layers.insert(0, key.meta.layer)
+        return state
+
+    @staticmethod
```

Comment:
There is a performance benefit to this, and seeing as it is going to be a common thing on battery power boards, I'll take everything that I can get. If you insist that it must go, I'll see to it, but I see no reason to not take battery and performance measures anywhere we can.

Reply:
I'm not aware of any performance benefits to this, other than *maybe* in the `mpy-cross` case, which isn't generally used right now. Decorators in Python are just functions that wrap other functions, `@staticmethod`s do not implicitly save CPU cycles by way of being attached to the class _definition_ instead of the _instance_.

It still stands that this is the one and only place in the extensions codebase, to my knowledge, that uses `@staticmethod` instead of instance variables.

The other thing I'll note, since I now had an excuse to read over this extension in totality, is that the fact that `@staticmethod` can even be used is a design problem. Extensions should store their state in `self`, not in `state` (which in other extensions was called `keyboard`, to make it clear that that was the global keyboard state). If a method on an extension doesn't require `self` to change the state of the board, that's concerning.

Are we able to reach a place where `Layers` is fully self-contained? Even if this involves some new methods added to the extension API (perhaps it's necessary to intercept matrix scans in a different way than the API currently allows, for example), I'd love to see `KMKKeyboard.active_layers` go away completely. If that's not possible, then we've proven that `Layers` isn't independent enough to be an `Extension`, and thus must stay part of core, much as tap dance currently must.

I guess I should have done this deep of a read on the first pass, but the ultimate verdict is that I consider addressing this thread, somehow, a merge-blocker.

Reply:
Layers maybe should fall back to core then. I will note that @staticmethod in extensions can be useful, such as HSV to RGB conversions are only useful within RGB. As for storing state in keyboard for an extension, this will also be useful to users. I have been working on an external extension (not core) that calculates WPM, and another extension that reads that live and displays it on an OLED extension. Unless there is a better way to pass data back and fourth, this will still remain the case. I'm fine with layers going into core, but I still feel that things will still break both of these "rules".

Reply:
I also intend to ultimately pull tap dance out, but leave all functions that make it work in core. Tapdange in it's current state is basically both "tapdance" as we understand it, as well as the only way to add/remove keys other than matrix events. I would like to decouple them and that would allow a nice distinction between an "extension" and "core". Extensions are something that allows you to manipulate the state, as well as extend it (layers is my prime example), while core offers you the bare functionality to get things working and helper functions to make extensions not repeat code as well as making it cleaner code.

Reply:
Cross-extension communication, as in the WPM<->OLED case, should be handled as a separate issue to this Layers case. In the WPM<->OLED case, there's a clear need to share data between extensions, and until/unless there is a specific API to handle that, the `Keyboard` object is the only place to stage data for handoff. In this case, however, the data is never, and should never be, shared, and should live fully within `self`. If it cannot, it's too tightly bound to core and should not be an extension.

Extensions that leave traces of themselves in core are by definitions not extensions - maybe they're modules, but they're not extensions. I cannot implement something like `Layers` without modifying core (and the underlying point here is of course that non-KMK-core developers also would not be easily able to implement something like `Layers` out-of-tree / in a private fork, and therefore IMO `Layers` is, again, not an extension at all). Thus, Tap Dance would also need to _fully_ remove itself from core to be an extension (and if core needs to add more callbacks to the plugin API to allow for that, then that is an engineering problem we should embrace in the name of correctness, and write the methods).

Reply:
Layers.py in it's current form is only able to manipulate layers and is possibly incorrectly named. Maybe it should be something like layer_manipulations. It is not what is intending to BE layers, only MANIPULATE the layers. There are many extension, or I guess "module" ideas that I can come up with that would manipulate key's, manipulate layers, and otherwise alter the state. It's possible that extensions should just be removed if this sort of use is beyond the scope of what it was it was intended, and replaced with a "module" placement. Most extensions that exist will alter the state of the keyboard directly, split and ble_split both come to mind. I will concede that layers in out of place if we also pull splits back into core, as these also massively alter the state and functionality of the core.